examples-gallery/intermediate/plot_countersteer.py

```diff
@@ -0,0 +1,279 @@
+# - :math:`v`: constant lontiguidnal speed at the rear contact
+# - :math:`g`: acceleration due to gravity
+# - :math:`m`: mass of bicycle and rider
+# - :math:`I_1`: roll principle moment of inertia
+# - :math:`I_2`: pitch principle moment of inertia
+# - :math:`I_3`: yaw principle moment of inertia
+h, a, b, v, g = sm.symbols('h a, b, v, g', real=True)
+m, I1, I2, I3 = sm.symbols('m, I1, I2, I3', real=True)
+
+# %%
+# The essential dynamics are described by a single degree of freedom model with
+# roll angle :math:`\theta` and its angular rate being the essential state
+# variables. The location of the rear contact in the ground plane :math:`x,y`
+# and heading :math:`\psi` will also be tracked. The input is the steering
+# angle :math:`\delta`.
+theta, thetadot = me.dynamicsymbols('theta, thetadot', real=True)
+x, y, psi = me.dynamicsymbols('x, y, psi', real=True)
+delta, deltadot = me.dynamicsymbols('delta, deltadot', real=True)
+t = me.dynamicsymbols._t
+dt = sm.symbols('Delta_t', real=True)
+
+# %%
+# The first two differential equations are the essential dynamics, followed by
+# the differential equatiosn to track :math:`x,y,\psi`. Noticing that both
+# :math:`\delta` and :math:`\dot{\delta}` are present as inputs to the model.
+# In the optimal control problem, both trajectories are saught with the
+# constraint that :math:`d\delta/dt` holds. To do this, add a differential
+# equation that ensures steering angle and steering rate variables are related
+# by differentiation and make :math:`\delta` a puesdo state variable with th
+# highest derivative :math:`\dot{\delta}` being the single unknown input
+# trajectory.
+eom = sm.Matrix([
+    theta.diff(t) - thetadot,
+    (I1 + m*h**2)*thetadot.diff(t) +
+    (I3 - I2 - m*h**2)*(v*sm.tan(delta)/b)**2*sm.sin(theta)*sm.cos(theta) -
+    m*g*h*sm.sin(theta) +
+    m*h*sm.cos(theta)*(a*v/b/sm.cos(delta)**2*deltadot +
+                       v**2/v*sm.tan(delta)),
+    x.diff(t) - v*sm.cos(psi),
+    y.diff(t) - v*sm.sin(psi),
+    psi.diff(t) - v/b*sm.tan(delta),
+    delta.diff(t) - deltadot,
+])
+MathJaxRepr(eom)
+
+# %%
+state_symbols = (theta, thetadot, x, y, psi, delta)
+MathJaxRepr(state_symbols)
+
+# %%
+# Provide some reasonably realistic values for the constants.
+par_map = {
+    I1: 9.2,  # kg m^2
+    I2: 11.0,  # kg m^2
+    I3: 2.8,  # kg m^2
+    a: 0.5,  # m
+    b: 1.0,  # m
+    g: 9.81,  # m/s^2
+    h: 1.0,  # m
+    m: 87.0,  # kg
+    v: 5.0,  # m/s
+}
+
+# %%
+# Instance constraints can be set on any of the state variables. The goal is to
+# transition from cruising at a steady state in balance equilibrium to steady
+# state in balance equlibrium with a 90 degree change in heading, i.e. make a
+# left turn.
+num_nodes = 201
+start = 0*dt
+end = (num_nodes - 1)*dt
+
+instance_constraints = (
+    # upright, no motion at t = 0
+    theta.func(0*dt),
+    thetadot.func(0*dt),
+    x.func(0*dt),
+    y.func(0*dt),
+    psi.func(0*dt),
+    delta.func(0*dt),
```

Comment:
Interestingly I had `0*h` here before (instead of `dt`) but the problem still solved.